### ISSUE TITLE
Another case of author component with an anonymous author

### ISSUE THREAD

**Issue by TysonMN** [CONNECT-platform/coding-blog-plugin#24]
My friend's blog post ([deployed](https://nathansweeney.coding.blog/dotnet-template) and [source](https://github.com/nathans1309/coding-blog/blob/master/posts/dotnet-template.md)) uses the `Author` component via `:Author src=github`.  I also think [their `github` configuration](https://github.com/nathans1309/coding-blog/blob/5d675dd7edb08ce7b70e57a9879b85cfb2deef80/.codedoc/config.ts#L36-L38) is correct (which was the issue in #18).  This is consistent when the output of `codedoc check`, which says
```
twilliams@EDMONDS MINGW64 /c/Code/ThirdParty/coding-blog (master)
$ codedoc check

# Checking CODEDOC configuration ...

# Checking Title configuration ...

# Checking GitHub configuration ...

# No issues detected.
```

I tried updating all dependencies to their latest version (by running `codedoc install` after deleting `node_modules` and `packages-lock.json`).  Still had the same behavior (when hosting locally).

What am I missing?  Why does the `Author` component on my friend's blog post show an anonymous author?

**Comment by loreanvictor** [comment]
Will check it later today, thanks for the heads up!

**Comment by loreanvictor** [comment]
on a quick look: this seems indeed like a bug with the `Author` component, since the commit data is correctly fetched from GitHub API on that page.

**Comment by loreanvictor** [comment]
@TysonMN @nathans1309 so this is the issue:

the author of the last commit (and the only commit on that file) is apparently NOT a GitHub user, so basically GitHub doesn't provide author information about them. This is the response from GitHub:

```json
[
  {
    "sha": "29acfcbed64dc4a04506b0259db17e625acad1cb",
    "node_id": "MDY6Q29tbWl0MjkxMTUzMzIwOjI5YWNmY2JlZDY0ZGM0YTA0NTA2YjAyNTlkYjE3ZTYyNWFjYWQxY2I=",
    "commit": {
      "author": {
        "name": "BTAA",
        "email": "nathan.sweeney@blocherconsulting.com",
        "date": "2020-09-08T21:18:05Z"
      },
      "committer": {
        "name": "BTAA",
        "email": "nathan.sweeney@blocherconsulting.com",
        "date": "2020-09-08T21:18:05Z"
      },
      "message": "first blog post",
      "tree": {
        "sha": "a899318408389622e28164c7e8a10492fce5d3a7",
        "url": "https://api.github.com/repos/nathans1309/coding-blog/git/trees/a899318408389622e28164c7e8a10492fce5d3a7"
      },
      "url": "https://api.github.com/repos/nathans1309/coding-blog/git/commits/29acfcbed64dc4a04506b0259db17e625acad1cb",
      "comment_count": 0,
      "verification": {
        "verified": false,
        "reason": "unsigned",
        "signature": null,
        "payload": null
      }
    },
    "url": "https://api.github.com/repos/nathans1309/coding-blog/commits/29acfcbed64dc4a04506b0259db17e625acad1cb",
    "html_url": "https://github.com/nathans1309/coding-blog/commit/29acfcbed64dc4a04506b0259db17e625acad1cb",
    "comments_url": "https://api.github.com/repos/nathans1309/coding-blog/commits/29acfcbed64dc4a04506b0259db17e625acad1cb/comments",
    "author": null,
    "committer": null,
    "parents": [
      {
        "sha": "9cc9cc78f3785540f6cff1a2de2c297815433e74",
        "url": "https://api.github.com/repos/nathans1309/coding-blog/commits/9cc9cc78f3785540f6cff1a2de2c297815433e74",
        "html_url": "https://github.com/nathans1309/coding-blog/commit/9cc9cc78f3785540f6cff1a2de2c297815433e74"
      }
    ]
  }
]
```

And as you can see the `author` object is `null`, and I also cannot click on the author name in GitHub to see their profile.

This of course is still a bug, since `Author` component can still deduce the name of the author and use that, but this would mean that the author displayed would not get an avatar and wouldn't be linked to anywhere in particular.

I will fix the issue with the component and roll an update when I get home.

**Comment by TysonMN** [comment]
Very good debugging @loreanvictor.

@nathans1309, your commit that added our first blog post uses your work email.  In contrast, the merge commit GitHub created for you when you completed [my PR](https://github.com/nathans1309/nathan-sweeney-blog/pull/1) uses your personal/gmail email.  I had this happen to me before.  You can [associate your work email with your GitHub account](https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account).  That will also fix this problem.

As a funny but related aside, since your only non-merge commit uses your work email (which is not associated with your GitHub account) and GitHub doesn't consider merge commits as a contribution, GitHub does not yet consider you a [contributor to your blog's repository](https://github.com/nathans1309/nathan-sweeney-blog/graphs/contributors).
![2020-09-22_11-50-57_303_chrome](https://user-images.githubusercontent.com/34664007/93913658-01624700-fccb-11ea-8d29-b568a9b51d7f.png)

**Comment by loreanvictor** [comment]
Fix deployed via `@codedoc/coding-blog-plugin@0.1.12`. As discussed before, the effect of this fix in the provided case would be that the name of the author (and publishing date) would be properly displayed, but not the author avatar or the url to their GitHub profile.

**Comment by TysonMN** [comment]
> You can associate your work email with your GitHub account . That will also fix this problem.

Another way that you can solve this problem @nathans1309 is by recreating that commit with you personal/gmail email and then force pushing.  Normally people recommend against force pushing, but this is your repo that almost no one else will ever use.

**Comment by TysonMN** [comment]
> ...the effect of this fix in the provided case would be that the name of the author (and publishing date) would be properly displayed, but not the author avatar or the url to their GitHub profile.

Why that fix?  I think it is possible to better with only a little extra work.

Given the username `MyUsername`, the URL to the corresponding GitHub profile is `https://github.com/MyUsername`, and [a URL to an image of the corresponding avatar](https://stackoverflow.com/questions/22932422/get-github-avatar-from-email-or-name/36380674#36380674) is `https://github.com/MyUsername.png`.

**Comment by loreanvictor** [comment]
@TysonMN which username are you referring to? In the provided case:
- the username on the commit is `BTAA`, however the url https://github.com/BTAA.png leads nowhere.
- the username of the repo can be used, but it would lead to two issues:
  - it is possible that the author of that particular post is indeed someone else without a publicly accessibly GitHub profile. In that case your proposed solution would lead to the incorrect avatar being fetched.
  - the proposed solution would introduce a coupling between the logic of `Author` component and a _convention_ (not a _specification_) of GitHub API, which by nature can be changed without prior notice, leading to errors.

**Comment by nathans1309** [comment]
> > You can associate your work email with your GitHub account . That will also fix this problem.
> 
> Another way that you can solve this problem @nathans1309 is by recreating that commit with you personal/gmail email and then force pushing. Normally people recommend against force pushing, but this is your repo that almost no one else will ever use.

Thanks, @TysonMN I made those changes.

**Comment by TysonMN** [comment]
> @TysonMN which username are you referring to?
> ...
> * the username of the repo can be used, but it would lead to two issues:
>   
>   * it is possible that the author of that particular post is indeed someone else without a publicly accessible GitHub profile. In that case your proposed solution would lead to the incorrect avatar being fetched.
>   * the proposed solution would introduce a coupling between the logic of `Author` component and a _convention_ (not a _specification_) of GitHub API, which by nature can be changed without prior notice, leading to errors.

I was referring to this case with the two issues that you mentioned.

The second issue is definitely a concern.  Maybe GitHub's API has a way to officially convert from a GitHub username to an avatar.  However, isn't this already a concern?  **runs away to check**  Oh, it is not an issue now because the `Author` component becomes a script that fetches this data from GitHub on page load.  I was guessing the data was obtained from GitHub at build time.

I hadn't considered the first issue.  Which commit are you assuming is the author?  The most recent commit that edited the file or first commit that created the file?

**Comment by loreanvictor** [comment]
> Oh, it is not an issue now because the Author component becomes a script that fetches this data from GitHub on page load. I was guessing the data was obtained from GitHub at build time.

Yeah thats precisely why its a client-side script, as it needs to be re-executed when the user navigates.

> I hadn't considered the first issue. Which commit are you assuming is the author? The most recent commit that edited the file or first commit that created the file?

Right now, the latest commit, since I suspected the last edit date should be displayed (so people can easily know if the post has been updated since the last time they saw it). This might not be ideal when multiple people start working on a post simultaneously though. My thinking was waiting for that to happen, and then act accordingly (I would personally love for it to display the first author and a limited number of most recent authors, with a link to GitHub so people can see its history for themselves, however depending on the case some other mechanism might be desirable).
